go/store/types/edits/async_sorted_edits.go

```diff
@@ -15,165 +15,93 @@
 // be merge sorted together.
 type AsyncSortedEdits struct {
-	sliceSize        int
-	sortConcurrency  int
-	asyncConcurrency int
-
-	ae         *atomicerr.AtomicError
-	sortChan   chan types.KVPSlice
-	resultChan chan types.KVPSlice
-	doneChan   chan bool
-	closed     bool
+	sliceSize       int
+	sortConcurrency int
+	closed          bool
 
 	accumulating []types.KVP
 	sortedColls  []*KVPCollection
 
 	nbf *types.NomsBinFormat
+
+	sortGroup *errgroup.Group
+	sortCtx   context.Context
+	sema      *semaphore.Weighted
 }
 
 // NewAsyncSortedEdits creates an AsyncSortedEdits object that creates batches of size 'sliceSize' and kicks off
 // 'asyncConcurrency' go routines for background sorting of batches.  The final Sort call is processed with
 // 'sortConcurrency' go routines
 func NewAsyncSortedEdits(nbf *types.NomsBinFormat, sliceSize, asyncConcurrency, sortConcurrency int) *AsyncSortedEdits {
-	ae := atomicerr.New()
-	sortChan := make(chan types.KVPSlice, asyncConcurrency*8)
-	resChan := make(chan types.KVPSlice, asyncConcurrency*8)
-	doneChan := make(chan bool, asyncConcurrency)
-
-	for i := 0; i < asyncConcurrency; i++ {
-		go func() {
-			defer func() {
-				doneChan <- true
-			}()
-
-			err := sorter(nbf, sortChan, resChan)
-			ae.SetIfError(err)
-		}()
-	}
-
+	group, groupCtx := errgroup.WithContext(context.TODO())
 	return &AsyncSortedEdits{
-		sliceSize:        sliceSize,
-		asyncConcurrency: asyncConcurrency,
-		sortConcurrency:  sortConcurrency,
-		ae:               ae,
-		sortChan:         sortChan,
-		resultChan:       resChan,
-		doneChan:         doneChan,
-		accumulating:     make([]types.KVP, 0, sliceSize),
-		sortedColls:      nil,
-		nbf:              nbf,
+		sliceSize:       sliceSize,
+		sortConcurrency: sortConcurrency,
+		accumulating:    make([]types.KVP, 0, sliceSize),
+		sortedColls:     nil,
+		nbf:             nbf,
+		sortGroup:       group,
+		sortCtx:         groupCtx,
+		sema:            semaphore.NewWeighted(int64(asyncConcurrency)),
 	}
 }
 
 // AddEdit adds an edit
 func (ase *AsyncSortedEdits) AddEdit(k types.LesserValuable, v types.Valuable) {
 	ase.accumulating = append(ase.accumulating, types.KVP{Key: k, Val: v})
-
 	if len(ase.accumulating) == ase.sliceSize {
-		ase.asyncSortAcc()
-	}
-}
-
-func (ase *AsyncSortedEdits) asyncSortAcc() {
-	ase.sortChan <- ase.accumulating
-	ase.accumulating = make([]types.KVP, 0, ase.sliceSize)
-	ase.pollSortedSlices()
-}
-
-func (ase *AsyncSortedEdits) pollSortedSlices() {
-	for {
-		select {
-		case val := <-ase.resultChan:
-			coll := NewKVPCollection(ase.nbf, val)
-			ase.sortedColls = append(ase.sortedColls, coll)
-
-		default:
+		coll := NewKVPCollection(ase.nbf, ase.accumulating)
+		ase.sortedColls = append(ase.sortedColls, coll)
+		toSort := ase.accumulating
+		ase.accumulating = make([]types.KVP, 0, ase.sliceSize)
+		if err := ase.sema.Acquire(ase.sortCtx, 1); err != nil {
```

Comment:
I could see this being a perf hit for large imports. It'll block the main thread and slow it from getting the data in.